pkg/mimir/tracing.go

```diff
@@ -117,7 +117,8 @@ func (t *OpenTelemetryTracerBridge) Start(ctx context.Context, spanName string,
 	}
 
 	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, t.tracer, spanName, mappedOptions...)
-	return ctx, NewOpenTelemetrySpanBridge(span, t.provider)
+	otelSpan := NewOpenTelemetrySpanBridge(span, t.provider)
+	return trace.ContextWithSpan(ctx, otelSpan), otelSpan
```

Comment:
Why you need otel span context originally? The change here would not hurt, since it only add otel span context key and value into context. Originally this opentelemetryTracerBridge is implemented so that all call to opentelemetry sdk would be translated to opentracing function behind. i don't think it is a bug, since it take assumption that since everything is using opentracing sdk behind, no otel context is needed.

Reply:
> why you think it might be a bug? 

It's a bug because `OpenTelemetryTracerBridge.Start` implements [`trace.Tracer.Start`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#Tracer), and the semantics of that method is to return a context containing a [`trace.Span`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#Span).

The bug happens when you call [`trace.SpanFromContext`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#SpanFromContext) within Prometheus, for example within [`tsdb.PostingsForMatchersCache`](https://github.com/grafana/mimir-prometheus/pull/553/files#diff-93f7c73af75f65a9cd81a56c7fa189dffe92c6943618cb9fb517360f0d277286R125). You then get a dummy (no-op) span back, instead of the real one.

Reply:
You can also see from my test that the original implementation is buggy, since `trace.SpanFromContext` won't work without my fix.